google-cloud-core/src/main/java/com/google/cloud/batchingexperimental/RequestAccumulator.java

```diff
@@ -0,0 +1,137 @@
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.batchingexperimental;
+
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Accumulates requests, used to batch many requests into one RPC call.
+ *
+ * <p>Experimental and only meant to be used by other packages in google-cloud-java.
+ *
+ * <p>Methods are not thread-safe.
+ *
+ * <p>Sample usage:
+ *
+ * <pre>{@code
+ * synchronized(accumulator) {
+ *   accumulator.add(item, future);
+ *   while (accumulator.hasBatch()) {
+ *     // If items and futures must be retained, they must be copied.
+ *     List<E> items = new ArrayList<>(accumulator.batch());
+ *     List<SettableApiFuture<R>> futures = new ArrayList<>(accumulator.futureBatch());
+ *
+ *     // ...
+ *
+ *     accumulator.next();
+ *   }
+ * }
+ * }</pre>
+ */
+@InternalApi
+public class RequestAccumulator<E, R> {
+  private final ArrayList<E> requests = new ArrayList<>();
+  private final ArrayList<SettableApiFuture<R>> futures = new ArrayList<>();
+  private long curBytes = 0;
+
+  // If not 0, the last element of requests is oversized and
+  // must be sent in its own batch.
+  private long oversizedBytes = 0;
+
+  private final int batchCount;
+  private final long batchBytes;
+
+  /**
+   * Creates a new accumulator, holding {@code batchCount} items and {@code batchBytes} bytes in a
+   * batch.
+   *
+   * <p>{@code batchBytes} is a soft limit. Once the current batch is larger than {@code
+   * batchBytes}, no new items will be added to the batch. {@code batchCount} is a hard limit. Once
+   * reached, no new items will be added.
+   */
+  public RequestAccumulator(int batchCount, long batchBytes) {
+    Preconditions.checkArgument(batchCount > 0, "batchCount must be positive");
+    Preconditions.checkArgument(batchBytes > 0, "batchBytes must be positive");
+    this.batchCount = batchCount;
+    this.batchBytes = batchBytes;
+  }
+
+  /**
+   * Add a new item {@code e} with size {@code bytes} into the current batch. The item is associated
+   * with a {@code future}.
+   *
+   * <p>Given a call {@code add(e, f)}, {@code e} and {@code f} are in the same index in {@link
+   * #batch()} and {@link #futureBatch()}.
+   *
+   * @throws IllegalArgumentException if {@code bytes} is not positive.
+   * @throws IllegalStateException if {@link #hasBatch()} would return true immediately before the
+   *     call.
+   */
+  public void add(E e, long bytes, SettableApiFuture<R> future) {
```

Comment:
Might be nicer to have this method create the future & return it

Reply:
I thought about this as well. My current thinking: logging doesn't require a future for every message, so it seems wasteful to create it unconditionally. The code required to create a future is also quite simple, so I don't feel *that* bad about making PubSub and BigTable write an extra one-liner.

If this turns out to be a great inconvenience, it's also easy to change later.

Do you feel strongly about this?

Reply:
I feel like the surface would be clearer if maybe we had a distinct method for the case where no future was involved.

Reply:
It'd still be equivalent to setting the future to null though right? Otherwise `batch()` and futureBatch()` won't be the same length anymore.